File: 39. Combination Sum Medium/solution.ts
function combinationSum(candidates: number[], target: number): number[][] {
    let result: number[][] = [];

    function backtrack(remain: number, combo: number[], start: number): void {
        if (remain === 0) {
            result.push(Array.from(combo));
            return;
        } else if (remain < 0) {
            return;
        }

        for (let i = start; i < candidates.length; i++) {
            combo.push(candidates[i]);
            backtrack(remain - candidates[i], combo, i);
            combo.pop();
        }
    }

    backtrack(target, [], 0);
    return result;
}

let candidates1 = [2, 3, 6, 7];
let target1 = 7;
console.log(combinationSum(candidates1, target1));
let candidates2 = [2, 3, 5];
let target2 = 8;
console.log(combinationSum(candidates2, target2));
let candidates3 = [2];
let target3 = 1;
console.log(combinationSum(candidates3, target3));

File: 39. Combination Sum Medium/solution.py
def combination_sum(candidates: list[int], target: int) -> list[list[int]]:
    result = []

    def backtrack(remain: int, combo: list[int], start: int):
        if remain == 0:
            result.append(list(combo))
            return
        elif remain < 0:
            return
        
        for i in range(start, len(candidates)):
            combo.append(candidates[i])
            backtrack(remain - candidates[i], combo, i)
            combo.pop()

    backtrack(target, [], 0)
    return result

candidates1 = [2, 3, 6, 7]
target1 = 7
print(combination_sum(candidates1, target1))
candidates2 = [2, 3, 5]
target2 = 8
print(combination_sum(candidates2, target2))
candidates3 = [2]
target3 = 1
print(combination_sum(candidates3, target3))